### PR TITLE
Ticket2994 keithley 2700

### PR DESCRIPTION
**Pull request by LiamPanchaud**
Note that this refactor is still in progress (on another branch) so that the emulator works with the VI. 

Therefore, the emulator still needs some changes and some tidying.

### CODE REVIEW

File: lewis_emulators/keithley_2700/device.py [hunk 1]
```diff
@@ -20,75 +33,107 @@ def _initialize_data(self):
...
-            self.channels.append(Channel(str(200+i)))
-        self.generate_channel_values()
-        self.fill_buffer()
+        self.channels = {}
+        for i in range(101, 110+1):
+            self.channels[str(i)] = Channel()
+        for i in range(201, 210+1):
+            self.channels[str(i)] = Channel()
 
     def generate_channel_values(self):
```
Comment (ghost) [inline lewis_emulators/keithley_2700/device.py]: Remove if unused

File: lewis_emulators/keithley_2700/device.py [hunk 2]
```diff
@@ -20,75 +33,107 @@ def _initialize_data(self):
...
+    def get_next_buffer_location(self):
+        if not self.is_buffer_full():
+            return len(self.buffer)
+        else:
+            return 0  # Buffer full so next loc is 0 after a clear
+
+    def is_buffer_full(self):
+        return len(self.buffer) >= self.buffer_size  # -1 because buffer is 0 indexed
 
     def fill_buffer(self):
```
Comment (ghost) [inline lewis_emulators/keithley_2700/device.py]: Remove if unused

File: lewis_emulators/keithley_2700/device.py [hunk 3]
```diff
@@ -1,15 +1,28 @@
...
 from lewis.core.logging import has_log
 from .states import DefaultState
 from lewis.devices import StateMachineDevice
 
+
+MAX_READ = 1500
+MIN_READ = 1000
+
+
 class Channel(object):
```
Comment (ghost) [inline lewis_emulators/keithley_2700/device.py]: Remove if unused

File: lewis_emulators/keithley_2700/device.py [hunk 4]
```diff
@@ -20,75 +33,107 @@ def _initialize_data(self):
...
-        self.data_elements = ""
-        self.auto_range_on = True
-        self.measurement_digits = 6
-        self.nplc = 5.0
-        self.scan_state_status = 2
-        self.scan_channel_start = 101
-        self.scan_channel_end = 210
-        self.buffer_count = 10
-        self.current_buffer_loc = 0
+        self.idn = "KEITHLEY"               # Device name
```
Comment (ghost) [inline lewis_emulators/keithley_2700/device.py:28]: You might need to check if there are old variables that are unused

File: lewis_emulators/keithley_2700/device.py [hunk 5]
```diff
@@ -20,75 +33,107 @@ def _initialize_data(self):
...
+        self.buffer_count = 10              # Number of samples to return. see $(P)COUNT in keithley_2700.db
 
         # Create channels from 101-110 and 201 - 210
-        self.channels = []
-        for i in range(1, 11):
-            self.channels.append(Channel(str(100+i)))
-            self.channels.append(Channel(str(200+i)))
-        self.generate_channel_values()
-        self.fill_buffer()
+        self.channels = {}
```
Comment (ghost) [inline lewis_emulators/keithley_2700/device.py]: remove if unused

File: lewis_emulators/keithley_2700/device.py [hunk 6]
```diff
@@ -20,75 +33,107 @@ def _initialize_data(self):
...
+                .encode("utf-8") for reading in self.buffer]
+
+    def clear_buffer(self):
+        """
+        Clears all buffer entries
+        """
+        self.buffer = []
+        self.log.info("=== Cleared Buffer ===")
 
     def set_channel_param(self, index, param, value):
```
Comment (ghost) [inline lewis_emulators/keithley_2700/device.py]: Remove set_channel/get_channel since unused

File: lewis_emulators/keithley_2700/device.py [hunk 7]
```diff
@@ -20,75 +33,107 @@ def _initialize_data(self):
...
+
+    def insert_mock_data(self, data):
+        """
+        Allows the insertion of specific, defined readings into the buffer
+        :param data: a list containing string representations of buffer readings
+        """
+        self.log.info("Inserting mock data into buffer: {}".format(data))
+        for item in data:
+            reading, timestamp, channel = item.split(",")
+            if not self.is_buffer_full():
```
Comment (ghost) [inline lewis_emulators/keithley_2700/device.py]: You don't need this if statement if the operation is pass. You can just have:

```python
            if self.is_buffer_full() and self.buffer_autoclear_on:
                self.clear_buffer()
            self.buffer.append(BufferReading(reading, timestamp, channel))
```

File: lewis_emulators/keithley_2700/interfaces/stream_interface.py [hunk 8]
```diff
@@ -10,68 +12,71 @@
...
-        CmdBuilder("set_auto_range_status", arg_sep="").escape("FRES:RANG:AUTO ").arg("OFF|ON").build(),
-        CmdBuilder("get_auto_range_status").escape("FRES:RANG:AUTO?").build(),
-        CmdBuilder("set_resistance_digits", arg_sep="").escape(":FRES:DIG ").int().escape(", (@").
-        int().escape(":").int().escape(")").build(),
-        CmdBuilder("set_resistance_rate").escape(":FRES:NPLC ").float().build(),
-        CmdBuilder("set_scan_state").escape("ROUT:SCAN:LSEL ").arg("INT|NONE").build(),
-        CmdBuilder("get_scan_state").escape("ROUT:SCAN:LSEL").build(),
-        CmdBuilder("set_scan_channels", arg_sep="").escape("ROUT:SCAN (@").int().escape(":").int().escape(")").build(),
-        CmdBuilder("get_scan_channels").escape("ROUT:SCAN?").build()
+        # get_multicmds splits commands by ';' if multiple command strings are received
```
Comment (ghost) [inline lewis_emulators/keithley_2700/interfaces/stream_interface.py:21]: Consider different formatting for these commands. I.e., 

````python
        CmdBuilder("set_measurement", ignore_case=True).escape(":FUNC '")
                                                       .arg("VOLT:DC|VOLT:AC|CURR:DC|CURR:AC|RES|FRES|CONT|FREQ|PER")
                                                       .escape("', (@").regex("[+-]?\d+").escape(":").regex("[+-]?\d+")
                                                       .escape(")").eos().build(),
````

Also, I believe the `ignore_case` argument is not required for all commands, as you've created these functions yourself.

File: lewis_emulators/keithley_2700/interfaces/stream_interface.py [hunk 9]
```diff
@@ -212,3 +216,23 @@ def get_scan_channels(self):
     def set_scan_channels(self, start, end):
         self._device.scan_channel_start = start
         self._device.scan_channel_end = end
+
+    def get_multicmds(self, command, other_commands):
```
Comment (ghost) [inline lewis_emulators/keithley_2700/interfaces/stream_interface.py:310]: The .proto file does not seem to be outputting multiple commands separated by semi-colons. If this is needed then it needs associated tests. If it's not used then maybe it doesn't need to exist.

File: lewis_emulators/keithley_2700/interfaces/stream_interface.py (hunk 8 above)
Comment (LiamPanchaud) [reply lewis_emulators/keithley_2700/interfaces/stream_interface.py:21]: `ignore_case` is needed to make the emulator compatible with the VI, which is a requirement of this project

File: lewis_emulators/keithley_2700/interfaces/stream_interface.py (hunk 9 above)
Comment (LiamPanchaud) [reply lewis_emulators/keithley_2700/interfaces/stream_interface.py:310]: This is again for VI compatibility and is covered by testing the emulator with the VI

File: lewis_emulators/keithley_2700/interfaces/stream_interface.py (hunk 8 above)
Comment (ghost) [reply lewis_emulators/keithley_2700/interfaces/stream_interface.py:21]: Thanks for the clarification.

**Pull request merged**